version: '3.7'

networks:
  network:
    driver: bridge

services:
  db:
    container_name: core_db
    image: postgres:16-alpine
    restart: always
    volumes:
      - ./db:/var/lib/postgresql/data
    env_file:
      - .env/.env
    networks:
      - network
    healthcheck:
        test: ["CMD-SHELL", "pg_isready -U $$POSTGRES_USER -d $$POSTGRES_DB -h db"]
        interval: 10s
        timeout: 5s
        retries: 5
    ports:
      - 5432:5432

  api:
    depends_on:
      db:
        condition: service_healthy
    container_name: core_api
    build:
      context: ./api
      dockerfile: api.Dockerfile
    restart: on-failure
    volumes:
      - ./api/app:/app
    env_file:
      - .env/.env
    networks:
      - network
    entrypoint: ["sh", "-C", "/scripts/start.sh"]

  app:
    depends_on:
      - api
    container_name: core_app
    build :
      context: ./app
      dockerfile: app.Dockerfile
    restart: on-failure
    volumes:
      - ./app/app:/app
      - ./.env/.env:/app/.env
    networks:
      - network
    entrypoint: ["sh", "-C", "/scripts/start.sh"]

  nginx:
    depends_on:
      - api
      - app
    container_name: proxy
    build:
      context: ./proxy
      dockerfile: proxy.Dockerfile
    restart: on-failure
    ports:
      - 80:80
      - 5111:5111
    volumes:
      - ./proxy/conf:/etc/nginx/conf.d
    networks:
      - network
    entrypoint: ["nginx", "-g", "daemon off;"]
